### PR TITLE
[eas-cli] integrate ascApiKey with submissions workflow

### PR DESCRIPTION
**Pull request by quinlanj**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.
- [x] I've tagged the changelog entry with `[EAS BUILD API]` if it's a part of a breaking change to EAS Build API (only possible when updating `@expo/eas-build-job` package).

# Why
App Specific Password - ASP
App Store Connect Api Key- ASC

This PR changes the workflow for Submission credentials to the following:
- If user specifies APP_SPECIFIC_PASSWORD env var, use ASP
- If user specifies ASC fields in `eas.json`, use ASC
- If user has setup ASC for their project in www, use ASC
- If we are sure user has compatible ASC in www, autosuggest to user. If they say yes, use ASC
- Else, give the user 3 options:
    - Use existing ASC in www. Only show this if user has ASC's.
    - Add new ASC. Either we generate for them, or user uploads.
    - **Enter ASP manually. We want to deprecate this option eventually.**

# How

For users who have always relied on manually inputting their ASP, they could get confused if we remove this option. They'd need to revisit the docs or look at our changelog to figure out the best way to proceed and if they just 'want the service to work', this could lead to them getting upset.

Probably a good idea to keep in the ASP prompt option for a while, and once ASC gets mass adoption, we can remove the prompt. Note that we will always support passing in ASP via env var.

# Test Plan

- [x] new tests pass

### CODE REVIEW

**Comment by quinlanj** [comment]
@dsokal @barthap should i also add some analytics to this? it's a lot of new code in a high traffic workflow, so might be a good idea

**Comment by quinlanj** [comment]
/changelog-entry new-feature Integrate ascApiKey with submissions workflow.

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ This is the log of notable changes to EAS CLI and related packages.
 - Allow users to assign an ASC Api Key to their project. ([#719](https://github.com/expo/eas-cli/pull/719) by [@quinlanj](https://github.com/quinlanj))
 - Add setup support for ASC Api Keys. ([#733](https://github.com/expo/eas-cli/pull/733) by [@quinlanj](https://github.com/quinlanj))
 - Show initiating user display name when selecting a build to submit. ([#730](https://github.com/expo/eas-cli/pull/730) by [@barthap](https://github.com/barthap))
+- Integrate ascApiKey with submissions workflow. ([#737](https://github.com/expo/eas-cli/pull/737) by [@quinlanj](https://github.com/quinlanj))
```
Comment (barthap) [inline CHANGELOG.md:16]:
```suggestion
- Integrate ASC Api Keys with submissions workflow. ([#737](https://github.com/expo/eas-cli/pull/737) by [@quinlanj](https://github.com/quinlanj))
```
to be consistent with other changelog entries

File: packages/eas-cli/src/submit/ios/IosSubmitter.ts [hunk 2]
```diff
@@ -152,6 +171,10 @@ function formatServiceAccountSummary({ summary }: AscApiKeyResult): string {
       label: 'Key ID',
       value: keyId,
     },
+    {
+      label: 'Key Name',
+      value: name,
+    },
```
Comment (barthap) [inline packages/eas-cli/src/submit/ios/IosSubmitter.ts:177]: Maybe we should display key name first, before other values - moving this to the top should do the job.

File: packages/eas-cli/src/submit/ios/__tests__/CredentialsServiceSource-test.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      profile: {
+        language: 'en-US',
+      },
+      nonInteractive: true,
+    });
+    jest
+      .spyOn(SetUpSubmissionCredentials.prototype, 'runAsync')
+      .mockImplementation(async _ctx => 'super secret');
+    asMock(getBundleIdentifierAsync).mockImplementation(() => 'com.hello.world');
+    const result = await getFromCredentialsServiceAsync(ctx);
```
Comment (barthap) [inline packages/eas-cli/src/submit/ios/__tests__/CredentialsServiceSource-test.ts:57]:
```suggestion

    const result = await getFromCredentialsServiceAsync(ctx);
```
nitpick: IMO one blank line between preparation and act/assertion part makes tests easier to read, but up to you.

**Review by barthap — APPROVED** [review]

File: packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import { CommonIosAppCredentialsFragment } from '../../../graphql/generated';
+import Log, { learnMore } from '../../../log';
+import { promptAsync } from '../../../prompts';
+import { CredentialsContext } from '../../context';
+import { UnsupportedCredentialsChoiceError } from '../../errors';
+import { AppLookupParams } from '../api/GraphqlClient';
+import { AppStoreApiKeyPurpose } from './AscApiKeyUtils';
+import { SetUpAscApiKey } from './SetUpAscApiKey';
+
+export const PROMPT_FOR_APP_SPECIFIC_PASSWORD = 'PROMPT_FOR_APP_SPECIFIC_PASSWORD';
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts:14]:
```suggestion
export const PROMPT_FOR_APP_SPECIFIC_PASSWORD = 'PROMPT_FOR_APP_SPECIFIC_PASSWORD';

```

File: packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+export class SetUpSubmissionCredentials {
+  private setupAscApiKeyAction;
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts:16]:
```suggestion
  private setupAscApiKeyAction;

```

File: packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  constructor(app: AppLookupParams) {
+    this.setupAscApiKeyAction = new SetUpAscApiKey(app, AppStoreApiKeyPurpose.SUBMISSION_SERVICE);
+    this.setupAscApiKeyAction.choices = this.setupAscApiKeyAction.choices.concat({
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts:19]: I know why you did it but modifying object members is a bad anti-pattern. Add a comment explaining why we need this.

File: packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      const { keyIdentifier, name } = nullthrows(
+        iosAppCredentials.appStoreConnectApiKeyForSubmissions,
+        'ASC Api Key must be defined for EAS Submit'
+      );
+      Log.log(`Using Api Key ID: ${keyIdentifier}${name ? ` (${name})` : ''}`);
+      return iosAppCredentials;
+    } catch (e) {
+      if (e instanceof UnsupportedCredentialsChoiceError) {
+        const choice = e.choice;
+        if (choice === PROMPT_FOR_APP_SPECIFIC_PASSWORD) {
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/actions/SetUpSubmissionCredentials.ts:39]:
```suggestion
        if (e.choice === PROMPT_FOR_APP_SPECIFIC_PASSWORD) {
```
or
```suggestion
        const { choice } = e;
        if (choice === PROMPT_FOR_APP_SPECIFIC_PASSWORD) {
```

File: packages/eas-cli/src/credentials/ios/actions/__tests__/SetUpSubmissionCredentials-test.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+} from '../../../__tests__/fixtures-ios';
+import { getAppLookupParamsFromContext } from '../BuildCredentialsUtils';
+import { SetUpAscApiKey } from '../SetUpAscApiKey';
+import {
+  PROMPT_FOR_APP_SPECIFIC_PASSWORD,
+  SetUpSubmissionCredentials,
+} from '../SetUpSubmissionCredentials';
+
+jest.mock('../../../../prompts');
+asMock(confirmAsync).mockImplementation(() => true);
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/actions/__tests__/SetUpSubmissionCredentials-test.ts:18]:
```suggestion
asMock(confirmAsync).mockImplementation(() => true);

```

File: packages/eas-cli/src/credentials/ios/actions/__tests__/SetUpSubmissionCredentials-test.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    jest
+      .spyOn(SetUpAscApiKey.prototype, 'runAsync')
+      .mockImplementation(async _ctx => testCommonIosAppCredentialsFragment);
+    const ctx = createCtxMock();
+    const appLookupParams = getAppLookupParamsFromContext(ctx, findApplicationTarget(testTargets));
+    const setupAscApiKeyAction = new SetUpSubmissionCredentials(appLookupParams);
+    const asc = await setupAscApiKeyAction.runAsync(ctx);
+
+    expect(asc).toEqual(testCommonIosAppCredentialsFragment);
+    jest.restoreAllMocks();
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/ios/actions/__tests__/SetUpSubmissionCredentials-test.ts:52]: This should be in the finally block or in `beforeEach`/`afterEach`.
```
try {
  // ... 
} finally {
  jest.restoreAllMocks();
}
```

File: packages/eas-cli/src/submit/ios/CredentialsServiceSource.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import { Platform } from '@expo/eas-build-job';
+import nullthrows from 'nullthrows';
+
+import { SetUpSubmissionCredentials } from '../../credentials/ios/actions/SetUpSubmissionCredentials';
+import Log from '../../log';
+import { getBundleIdentifierAsync } from '../../project/ios/bundleIdentifier';
+import { findAccountByName } from '../../user/Account';
+import { SubmissionContext } from '../context';
+import { AscApiKeyResult } from './AscApiKeySource';
```
Comment (dsokal) [inline packages/eas-cli/src/submit/ios/CredentialsServiceSource.ts:9]:
```suggestion
import { AscApiKeyResult } from './AscApiKeySource';

```

**Review by dsokal — APPROVED** [review]

**Comment by dsokal** [comment]
> should i also add some analytics to this? it's a lot of new code in a high traffic workflow, so might be a good idea

I think it's always better to have more metrics than not to have them at all. So yeah, go ahead.

**Pull request merged**